### DISCUSSION TITLE
Feedback

### DISCUSSION THREAD

**Discussion by cevr** [bvaughn/suspense#23] [Ideas]
Here's a little app i made for demonstrative purposes: [stackblitz.com/edit/vitejs-vite-v6umf3?file=src/App.tsx](https://stackblitz.com/edit/vitejs-vite-v6umf3?file=src/App.tsx)

After using this library for a couple days, I've gathered some feedback:

Will make a comment for each

**Comment by cevr** [comment]
Possible to add an invalidate function to the cache?

The implementation I use is:
```ts
function invalidate<TParams extends any[], TValue>(
 cache: Cache<TParams, TValue>,
 ...args: TParams
) {
 cache.evict(...args);
 cache.prefetch(...args);
}
```

**Reply by bvaughn** [reply]
Initial thoughts: This doesn't seem too useful, if it's literally just wrapping those two methods.

The "useful" thing would be if it also scheduled an update with React– but doing that requires either using a hook (aka [`useCacheMutation`](https://suspense.vercel.app/useCacheMutation)) or manually passing in a "refresh" function from `useCacheRefresh` (which feels too awkward to be useful IMO).

**Reply by cevr** [reply]
After thinking about this, I think one thing that is missing from the mutation story is that it expects the result of the mutation to also have the new value for whatever cache. From my experience, it is often that APIs just return some sort of success indicator rather than the updated value. On top of that, unless there is a normalized cache that is able to read the underlying type and schema of the API, it fails to account for other caches that reference that same value. eg: An update to the UserCache does not update UsersCache etc.

There's also the consideration of maybe supporting optimistic updates. Where you supply a sync value of the expected result after the mutation and otherwise rollback if the mutation fails.

This may be why it's not considered too useful for now? In cases like these, calling an invalidate method would be very common

**Reply by bvaughn** [reply]
Admittedly the current mutation hook is a first pass at supporting mutation 😄  I think the pattern could be expanded.

For instance, "optimistic updates" _could_ be built on top of the mutation hook by combining a sync and async mutation. Sketched out, it might look something like:
```js
const { isPending, mutateAsync, mutateSync } = useCacheMutation(cache);

const onClick = () => {
  mutateSync(params, nextValue);

  mutateAsync(params, async () => {
    const success = await api(...);
    return success ? nextValue : prevValue;
  });
};

// Use isPending to disable UI during the update
```

> An update to the UserCache does not update UsersCache etc.

I think data normalization is a bit larger scope than I'm ready to take on at the moment. You're right though that this package doesn't handle that at all for now.

**Reply by cevr** [reply]
I agree that data normalization is whole other beast! I like the invalidate function since it follows the pull model more closely.

I also like your example, it's where I would say that it would be helpful. If the mutation is a success, call invalidate for another cache (or caches) while setting the next value for the cache provided in the `useCacheMutation`

**Comment by cevr** [comment]
Is it possible to notify subscribers when an eviction happens? Currently no notification is given, unless its an evictAll i believe.

similarly, evicting and forcing an update afterwards does not cause the suspense fallback to show. Intended? You can see it in the app above by going to the posts, and clicking evict current page

**Reply by bvaughn** [reply]
Ah, yes. This looks like a bug. I've added unit tests for it and fixed it– ab461f1

**Comment by cevr** [comment]
The current eviction policy seems unpredictable and aggressive, sometimes causing surprising evictions. In the app above, if you go to the posts, you'll see that sometimes when pressing the next button, a complete cache blowout has occurred forcing a suspense fallback. Will also notice it in the users demo (loading a new user will sometimes show all other previously loaded users are now evicted).

I was thinking a possible strategy is to allow the user to provide their own implementation of the internal cacher so to leave the eviction policy to them. I would personally want to use a LRU cache, but some might consider another type of cache. Providing the inversion of control could satisfy all use cases. What do you think?

**Reply by bvaughn** [reply]
I've considered an LRU option (and am still open to that in the future)– but it sounds like you're just wanting to disable weak refs? That can already be done via an option: https://suspense.vercel.app/guide/memory-management

**Reply by cevr** [reply]
I think that even if you had the weak refs, it seems oddly aggressive to a point of being detrimental to the user experience. Of the choice between both, I think an LRU allows for a much broader scope of customization on policy than weak refs.

**Reply by bvaughn** [reply]
I hear what you're saying. This is something I think I'd like to sit on for a bit.

You _could_ disable weak refs and wrap a cache with an LRU to accomplish what you're aiming for, but admittedly that's kind of clunky. I'm concerned it might be complicated for me to support a generic "cache" injection though– because types like map/object/LRU have pretty different semantics than weak ref. Essentially I need to handle weak ref semantics at a lower level, which makes it harder to accept an external param like you're talking about.

**Reply by cevr** [reply]
Yeah it's definitely a hard problem. My thoughts are like this: if I can't control the eviction policy to some degree, then I won't use it and in this case means disabling the weak refs. At least with an LRU you have many options to use to decide how you want your cache evicted. A generic cache would be an even harder problem no doubt haha

**Reply by bvaughn** [reply]
Yeah, I hear that. I think I _won't_ do anything on this for now beyond just thinking about it, although I would be open to reviewing a PR if you wanted to put one together.

**Reply by cevr** [reply]
Okay! I'll research to see what a generic cache would look like, would be a learning experience

**Comment by cevr** [comment]
I commonly want to know if the current cache item is revalidating (no suspense fallback). What do you think about an API like this:
```ts
	const [value, revalidating] = Cache.read()
```
my current implementation for this is using this helper function:

```ts
function useRead<TParams extends any[], TValue>(
  cache: Cache<TParams, TValue>,
  ...args: TParams
) {
  const value = cache.read(...args);
  const status = useCacheStatus(cache, ...args);
  return [value, status === 'pending'] as const;
}
```
which seems to work

**Reply by bvaughn** [reply]
I don't think I understand the use case here enough to know if I think it would be widely useful enough to add an API for. (Particularly if you're able to accomplish it with only a few lines of code.) Can you elaborate?

**Reply by cevr** [reply]
Yeah I think its fine to be honest, a simple preference that can be solved rather easily as you said

**Comment by cevr** [comment]
provide built in hooks? this one is not really a big deal and philosophically i can understand it not being part of the cache. But something like `useCacheStatus(cache, ...)` to `Cache.useStatus(...)` etc

**Reply by bvaughn** [reply]
I thought about this but it felt a little bloaty (subjective). Something I'll think on a bit more though.

**Reply by cevr** [reply]
Yeah I agree its very subjective.

**Comment by cevr** [comment]
the current getValueIfCached throws an error if rejected, which makes it cumbersome in some cases. Thoughts on providing a `Cache.peek(...)` function that returns the raw record or undefined?

**Reply by cevr** [reply]
In my implementation i paired it with the function `cache.evictIf(predicate, ...args)` that was useful in ErrorBoundaries to evict if the cache was rejected before reseting the error boundary

**Reply by bvaughn** [reply]
I don't want to expose the raw `Record` because I want to preserve the option of changing that structure (like I did in e647dfd9b48fc71d66ff4f0b53dcf0e1297a627b).

It sounds like you're almost looking for the value returned by [`useImperativeCacheValue`](https://suspense.vercel.app/useImperativeCacheValue), but maybe without the side effect of kicking off a fetch?

> Thoughts on providing a `Cache.peek(...)`

So more to the point– I'm not sure I see the value of this over the current API, which lets you imperatively ask if the record is _errored_ (`cache.getStatus`) before asking for its value. (You could write your own "peek" method pretty easily.)

```js
function peek(cache, ...params) {
  switch (cache.getStatus(...params)) {
    case STATUS_RESOLVED:
      return cache.getValue(...params);
    default:
      return undefined;
  }
}
```

**Reply by cevr** [reply]
Fair enough, I think that makes a lot of sense. As you said, creating one's own peek is simple enough using the `getStatus`

**Comment by cevr** [comment]
the `subscribeToStatus` calls the subscription callback immediately causing an unnecessary render on mount when using uSES. Is this intended?

**Reply by bvaughn** [reply]
Hm, it _was_ intentional. I wouldn't have expected it to cause an unnecessary React render though. I would have expected React to check the `getValue` callback, see that it hadn't changed, and bail out without rendering.

**Reply by cevr** [reply]
Yeah those were my thoughts as well but it doesn't seem to be the case. overall not a huge problem either way

**Reply by bvaughn** [reply]
Are you perhaps just observing [strict mode / strict effects](https://reactjs.org/docs/strict-mode.html) behavior (DEV only)?

**Reply by cevr** [reply]
🤦 you're right

**Reply by bvaughn** [reply]
Haha, no worries! It gets a lot of people.

**Comment by cevr** [comment]
the type inference feels lacking, I think the `load` function should provide all the signatures it needs instead of having to manually supply the generics. Thoughts?

**Reply by bvaughn** [reply]
> the type inference feels lacking

This is kind of vague. Show me what you're talking about? Admittedly, I'm still kind of new to TypeScript. I would be happy if there was a way to improve the type ergonomics without sacrificing safety.

**Reply by cevr** [reply]
Let me cook up an example

**Reply by cevr** [reply]
https://stackblitz.com/edit/react-ts-8aqqze?file=App.tsx

Here's an example, the SuspenseCache class automatically infers the TParams and TValue based on the `load` function provided

**Reply by bvaughn** [reply]
I _think_ my current API can be relaxed without losing the type hints?

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMKIBYQDwAU5RwC2AzshAB6QgAmZAglIQJ7ZwjMB8ANMgGpwANgFcInZAG8AUMmQJMEABQA3IaIBc-NRF4A6fQAcCxEpvyFSASk3KA9sBoBuGcigQ4NeiWYgEi-bpGFqbI5ibWoVC2RMAkEAAywADWOAIiYsgAPlrpzgC+UlJgzAYo6AhY8bYeAPIGYMC2IGQAvJIuJMAA5iBCmvQARrZQYADK3b2C+c5FJWVucJDlWHUNTSR4xqTkVBC0DExwrOxcvGmi4m3Ssl0QYADSEMwA-Jr+hlshYVbILeIkYCgoC6zlkgmqNDeLlkASCJk0AG0At8SLxlgkIatGs0ALouSy-cS4KIxOKJFLYc4ZbJU6aFSgGYZgZAwYS+NYgOQLJYKTbBHbUOjIRgsNgcHg5C6KFy2erYkKobkQdFY9Z8kxnbScKQRdHq0ia9Lia5ydbMiTIW4PJ68cEeZB5X7IWUckgzWRuMDCKCck2yeQVJSqdKaKl6D7BMyfCJ2BztWQJ5AAeiTyAC0Id3AzC083l870CnyjwQixOisQSyVSWqykpQfoTnu9nLtNALcINkk6PT6yBAEAA7sKhiNUE1AbZBIJoIpLLpu5M8pZQQm8lnZHl8oUEGbkACgSAuuinQglejFCarY9mG8QMIiJo70QBtBePvNPvgQS-nvAevkK2mhwHmCDIIoT6PveL5QG+gIfoCX6EvGjZ3M2v5QCuBRLjMKbIDQtgQGQIC2My0BRFAUifoe6K6AGWCKAA5PuDG8AAjAATAAzLwDEDIIcAYAxy6FLhZHDCQlEIdRCi0QoiiAqIvAKToyDKcJ267rYAwAFYQAgYDHm0p7uDygYXi4V5PG8BAsJBz7QAiOK8EMk6aC507sN+4gAAYACQSDZRy6NOh5gBgeTqP57l5N5-6AcgwE+KBiiBTefZQQ5TnIO5bm2JO7ggF5yEeqhPqSNleWCJoykOphUjYSJqb4YRfYkeQTDDFIWm6fpNF0UoEg5cpa7IAirFOWpOGpmJUASd1ekGTJ-UXjlDFwAMCAMSNY0TVAojCfNvVLXJg2Veow28AARMBNCXdwk1dTpC19SdQ17RA23jdwrEHU9R2BrJZl3lOvA7fd72-T1i0A8tylKe98P7Y4QA

**Reply by cevr** [reply]
yeah that's perfect! my only real want is not being _forced_ to explicitly define the requirements of load

**Reply by bvaughn** [reply]
Have you tried just _not_ doing that and seeing if it works? I think the only reason your example didn't work is that the function omitted the "options" param (which confused the generics)

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYQIYGMAWBTAMhVAEwHkwZgIA7AZwBo50ptUZs0ts4BfOAMyggg4AcmoBXamGw1swgNwAoBeirV4Ad1TB4AXjgAKENQBccSmJAAjbFACUcHQD4z2dXAAKAkMGrZ9+pmoIABsAN2x7JzhfGAAVUGwIMRgA7CCw7HojW1tFBVRqAE9KdD4xEvIqPmwYLABVXyh9YEJTNShgSgBzeggyChpTdhx8IlJKmnsEBTg4VE1tOAWUgEYABg3cmbgmGDEoSkRt2Zbaba5FLiUVGngGm2HOPUZmVkf9adngglbq2sx7lAzlwtgoAPRguA2ARQagKQGPAB0TCI+lBELghAgaUo8Gh0HhjSRKMI+gARAVCGTckA

**Reply by cevr** [reply]
Ah I see. If the `CacheLoadOptions` are not intended optional in the `load` function, maybe something like this could convey that?

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMKIBYQDwAU5RwC2AzshAB6QgAmZAglIQJ7ZwjMB8ANMgGpwANgFcInZAG8AUMmQJMEABQA3IaIBc-NRF4A6fQAcCxEpvyFSASk3KA9sBoBuGcigQ4NeiWYgEi-bpGFqbI5ibWoVC2RMAkEAAywADWOAIiYsgAPlrpzgC+UlJgzAYo6AhY8bYeAPIGYMC2IGQAvJIuJMAA5iBCmvQARrZQYADK3b2C+c6FRSVlbnCQ5Vh1DU0keMak5FQQtAxMcKzsXLxpouJt0rJdEGAA0hDMAPya-obbIWFWyC3iJDAUFAXWcskE1RommwVQ8PzIlGodGQAG0AvDeCsEpC1o1mgBdTiKFyyAJBEyaWE0eG7JFkNH6DFoBRU3EbfHIF6o9FfTEsnH1PEkDmaEAQZTQFyWP7iXBRGJxRIpbAXDLZVXTQqUAzDMDIGDCXzrEByRbLBRbYK0-bIxgsNgcHg5S7E2S2QUbTSoM0QLFs5qWkznbScKQRLGB0jB9LiG5yDZ6iTIO6PZ68CEeZB5P7Id3GkgzWRuMDCKAmuOyeQVJSqdKaVV6T7BMxfCJ2BztWRd5AAeh7yACJKz3CHi083l8H0CXxbwQicuisQSyVSIayzpQFa7xdLJozNCn5Kjkk6PT6yDFAHdkINdagmkDbIJBNBFJZdKfJnlLGCu3kR7IeT5IUCAJsggLAiAXRYjmCA+liihximTzMO8IDCEQooYQM0C8BBmgQSC0r-OBQIAcg+6aHAE4IMgijoZhF7YbhpFQARQJETKnbbvcu6sb+BTfjMfbIDQtgQGQIC2Hq0BRFAUiEVBWK6FWWCKAA5BB6m8AAjAATAAzLw6kDIIcAYOpP6FCJsnDCQCkcUpCgqQoihAqIvDuToyBeVZUjUT4tEGkaeIUZCABCthPooBAsFhRA4VAKL4rwQxPpoaUvuw0rcUWvFlpIyCZZoXlZgJsygc0eq2AMABWEAIGAMFtHB7jmtWiEuMhzzvLFxzxYlyWpVFggZSN7ggMR4gAAYACQSH1zC6C+UFgBgeTqPNmV5NN5GUWFHiRU+UhCdZ-ZiRJF7SeQTDDFINX1Y1ymqUoEjFV5-6ojpKW+cJ-a2VA9kPQ1TXOS9iHFepcADAg6mfSi33cL9911SDz2uW9I3qB9vAAETUTQuNI1AohWcDT1gxj70kxA8OIzpZOoxT1YuR16HPrwCM-TTjOPaDLPg15nk08LpOOEAA

**Reply by bvaughn** [reply]
I think it's trickier than that. It _is_ kind of optional. (Not all request types can be aborted.)

**Reply by cevr** [reply]
Then I think the relaxed one is a great middle ground, though I think putting CacheLoadOptions as the first argument might be better for discoverability. If the request can't be aborted, the first argument can just be ignored

**Reply by bvaughn** [reply]
I think first vs last parameter is subjective. I prefer it to be last in this case, but I'll give it a bit more thought :smile:

**Comment by cevr** [comment]
`not-started` seems like an odd status. I believe it means something like `cache-miss` unless you can set a cache record with `not-started`? If not, I think `undefined` could be a better descriptor of `cache-miss`

**Reply by bvaughn** [reply]
I dig this.

**Reply by bvaughn** [reply]
I wonder if "not-found" would be better, semantically?

**Reply by bvaughn** [reply]
Renamed! befed09

**Reply by cevr** [reply]
Even better! :D
